pkg/tikvclient/client.go

```diff
@@ -0,0 +1,146 @@
+package tikvclient
+
+import (
+	"context"
+	"crypto/tls"
+	"sync"
+
+	"github.com/pingcap/kvproto/pkg/tikvpb"
+	"github.com/pkg/errors"
+	"github.com/tikv/pd/pkg/utils/grpcutil"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/connectivity"
+)
+
+type clientConn struct {
+	// The target host
+	target string
+	conn   *grpc.ClientConn
+}
+
+// Close gprc connection.
+func (c *clientConn) Close() error {
+	if c.conn != nil {
+		return c.conn.Close()
+	}
+	return nil
+}
+
+// Get grpc.ClientConn.
+func (c *clientConn) Get() *grpc.ClientConn {
+	return c.conn
+}
+
+func newClientConn(ctx context.Context, addr string, tlsConfig *tls.Config, grpcDialOpt []grpc.DialOption) (*clientConn, error) {
+	c, err := grpcutil.GetClientConn(ctx, addr, tlsConfig, grpcDialOpt...)
+	if err != nil {
+		return nil, err
+	}
+	return &clientConn{
+		target: addr,
+		conn:   c,
+	}, nil
+}
+
+type option struct {
+	gRPCDialOptions []grpc.DialOption
+	tlsConfig       grpcutil.TLSConfig
+}
+
+// Opt is the option for the client.
+type Opt func(*option)
+
+// WithSecurity is used to set the TLSConfig.
+func WithSecurity(tlsConfig grpcutil.TLSConfig) Opt {
+	return func(c *option) {
+		c.tlsConfig = tlsConfig
+	}
+}
+
+// WithGRPCDialOptions is used to set the grpc.DialOption.
+func WithGRPCDialOptions(grpcDialOptions ...grpc.DialOption) Opt {
+	return func(c *option) {
+		c.gRPCDialOptions = grpcDialOptions
+	}
+}
+
+// RPCClient is RPC client struct used to manages connections with tikv-servers.
+type RPCClient struct {
+	sync.RWMutex
+
+	clientConns map[string]*clientConn
+
+	isClosed bool
+
+	option *option
+}
+
+// NewRPCClient creates a RPC client
+func NewRPCClient(opts ...Opt) *RPCClient {
+	cli := &RPCClient{
+		clientConns: make(map[string]*clientConn),
+		option:      new(option),
+		isClosed:    false,
+	}
+	for _, opt := range opts {
+		opt(cli.option)
+	}
+	return cli
+}
+
+// Close closes all connections.
+func (r *RPCClient) Close() {
+	r.Lock()
+	defer r.Unlock()
+	if !r.isClosed {
+		r.isClosed = true
+		for _, cc := range r.clientConns {
+			cc.Close()
+		}
+	}
+}
+
+func (r *RPCClient) getClientConn(ctx context.Context, addr string) (*clientConn, error) {
+	r.Lock()
+	defer r.Unlock()
+
+	if r.isClosed {
+		return nil, errors.Errorf("rpcClient is closed")
+	}
+	cc, ok := r.clientConns[addr]
```

Comment:
Do we need to make the connection long-alive? gc request may call rarely? I wonder if there are too many tikvs.

Reply:
Long-alive is ok, because if there is no request for 5 minutes, the connection becomes [idle](https://grpc.github.io/grpc/core/md_doc_connectivity-semantics-and-api.html). The connection long-alive also can make `tikv client` more general.